Documentation/design-docs/DotNetCore-SharedPackageStore.md

```diff
@@ -0,0 +1,218 @@
+
+# Building apps with shared packages
+
+The current mechanism to build applications that share assemblies is by not specifying a RID in the project file. Then, a portable app model is assumed and assemblies that are part of Microsoft.NETCore.App are found under the `dotnet` install root. With shared package store, applications have the ability to filter any set of packages from their publish output. Thus the decision of a portable or a standalone application is not made at the time of project authoring but is instead done at publish time. 
+
+## Project Authoring
+We will by default treat `Microsoft.NETCore.App` as though `type: platform` is always specified, thus requiring no explicit RID specification by the user. It will be an `ERROR` to specify a RID in the csproj file using the `<RuntimeIdentifier/>` tag.
+
+## dotnet restore
+
+Because RIDs are not available until publish time, the application is treated as though it is a present day portable app and a regular restore is performed.
+
+## dotnet build
+
+`dotnet build` should treat any project as though it is a portable app and produce `runtimeconfig.json` with the `Microsoft.NETCore.App` as the framework. In addition, the `runtimeconfig.json` file should also specify the TFM field as:
+
+```json
+"runtimeOptions": {
+    "tfm": "netcoreapp2.0",
+    "framework": {
+        "version": "2.0.0",
+        "name": "Microsoft.NETCore.App"
+    }
+}
+```
+Note that this is different from current behavior of `dotnet run` for an application that specifies the `<RuntimeIdentifier/>` tag in the csproj.
+**Current Behavior:** Picks `M.N.A` assemblies out of the NuGet cache without taking advantage of optimizations available from the shared `Microsoft.NETCore.App`.
+**New Behavior:** Picks `M.N.A` assemblies from the shared framework and the rest of them from the shared package store or the NuGet cache.
+
+`dotnet build` can take advantage of the `refs` folder available at the `packages/install/` folder from the `dotnet` root directory enabling the offline-restore-build scenario. Although we are designing to augment `dotnet build` in the future regarding the reference assemblies, for the scope of this work we'll focus only on runtime assemblies.
+
+## Host probe precedence
+
+The host will probe in the following order of precedence for `dotnet run` and application activations post `dotnet publish`:
+
++ `$CORE_SERVICING` on Unix or `%ProgramFiles(x86)%\coreservicing` on Windows.
++ `DOTNET_SHARED_PACKAGES_ROOTS` in the chained order
++ The user local shared package store in `%USERPROFILE%\.dotnet\packages` or `$HOME/.dotnet/`
++ The global shared package store
++ The Shared Framework directory
++ ~~`DOTNET_HOSTING_OPTIMIZATION_CACHE` deprecated in favor of `DOTNET_SHARED_PACKAGES_ROOTS`~~
++ `--additionalprobingpaths` specified in the command line
++ `runtimeOptions.additionalProbingPaths` (includes NuGet cache probe specified by the CLI for `dotnet run`)
++ Application `bin` directory
```

Comment:
Presently the Application is the last place we check for the assemblies. 
1. Should we provide a mechanism to opt out of the all the above locations ?
2. Provide a mechanism for app to win over others ?